rust-src/wallet_library/src/statement.rs

```diff
@@ -0,0 +1,375 @@
+    pub attribute_check: AttributeCheck<'a, TagType, AttributeType>,
+    pub _marker:         PhantomData<C>,
+}
+
+pub trait AttributeTagType: Serialize + Ord + Hash + Display {}
+impl<T: Serialize + Ord + Hash + Display> AttributeTagType for T {}
+
+type AttributeCheck<'a, TagType, AttributeType> =
+    Box<dyn Fn(&TagType, &AttributeType) -> AcceptableRequestResult + 'a>;
+
+#[derive(Debug, thiserror::Error, Clone)]
+pub enum RequestCheckError {
+    #[error("Credential statement must include atomic statements")]
+    EmptyCredentialStatement,
+    #[error("Credential statement may not reuse attribute tags across atomic statements")]
+    DuplicateTag,
+    #[error("Web3Id statement support have not been added yet")]
+    Web3IdStatementNotSupported,
+    #[error("`{0}`")]
+    InvalidValue(String),
+    #[error("Range statement min must be less than max")]
+    RangeMinMaxError,
+    #[error("The tag `{0}` is not allowed to be used for range statements")]
+    IllegalRangeTag(String),
+    #[error("Membership and NonMembership statement's may not have empty sets")]
+    EmptySet,
+    #[error("The tag `{0}` is not allowed to be used for set statements")]
+    IllegalSetTag(String),
+}
+
+impl<C: Curve, AttributeType: Attribute<C::Scalar>> AcceptableRequest<C, AttributeType>
+    for Request<C, AttributeType>
+{
+    fn acceptable_request(
+        &self,
+        config: &WalletConfig<C, AttributeType>,
+    ) -> AcceptableRequestResult {
+        self.credential_statements
+            .iter()
+            .try_for_each(|s| s.acceptable_request(config))
+    }
+}
+
+impl<C: Curve, AttributeType: Attribute<C::Scalar>> AcceptableRequest<C, AttributeType>
+    for CredentialStatement<C, AttributeType>
+{
+    fn acceptable_request(
+        &self,
+        config: &WalletConfig<C, AttributeType>,
+    ) -> AcceptableRequestResult {
+        match self {
+            CredentialStatement::Account {
+                statement,
+                network: _,
+                cred_id: _,
+            } => {
+                if statement.is_empty() {
+                    return Err(RequestCheckError::EmptyCredentialStatement);
+                }
+                let mut used_tags = HashSet::<AttributeTag>::new();
+                for atomic_statement in statement {
+                    let attribute = atomic_statement.attribute();
+                    if used_tags.contains(&attribute) {
+                        return Err(RequestCheckError::DuplicateTag);
+                    }
+                    used_tags.insert(attribute);
+
+                    atomic_statement.acceptable_atomic_statement(&config.identity_rules)?
+                }
+                Ok(())
+            }
+            CredentialStatement::Web3Id {
+                ty: _,
+                network: _,
+                contract: _,
+                credential: _,
+                statement: _,
+            } => Err(RequestCheckError::Web3IdStatementNotSupported),
```

Comment:
I'm a bit confused with this. There is a web3_rules in the walletconfig but you don't use them here at all.

Reply:
Yeah, I meant to implement this (and remove that error). I've added it.